Add installer build/test to the runtime.yml pipeline

dagood opened the pull request:
Adds installer build and test to `runtime.yml`, consuming live artifacts. Preserves the old standalone installer build + test pipeline using cached bits for now, but if it breaks with https://github.com/dotnet/runtime/pull/494 as expected I won't try to repair it with this PR.

Old WIP PR: https://github.com/dotnet/runtime/pull/529.

/cc @trylek @safern @jashook

jkoritzinsky committed:
First pass to get an installer build integrated into runtime.yml with minimal changes to the installer jobs.

Fix paths to jobs.

Fix typos

Add platform to parameter lists.

Pass platform to all jobs.

Cleanup variable usages.

Don't pass buildConfig down to installer jobs. They use matrix strategy instead.

Fix variable usage in windows build.

Fix windows platform names.

Add missing platform parameter for Windows_NT_arm

Update last condition re parameters.dockerImage.

Rename dockerImage parameter to productBuildDockerImage to be more explicit.

Use pool and container parameters from platform-matrix in installer build.

Pass rootfs dir via crossrootfsDir to bash-build.yml.

Fix passing down container in runtime-installer pipeline.

Remove unused Linux_x64_raw config.

jkoritzinsky committed:
Add support for clang 9 to corehost build.

jkoritzinsky committed:
Skip tests on Linux_musl_arm64

dagood committed:
runtime.yml installer build with live artifacts

Includes refactoring the runtime legs to avoid duplicating artifact
transfer logic and preserve the runtime standalone build.

dagood committed:
Fix installer overrides: accept live CI artifacts

safern commented on eng/pipelines/common/platform-matrix.yml:
```diff
@@ -229,6 +231,7 @@ jobs:
       helixQueuesTemplate: ${{ parameters.helixQueuesTemplate }}
       osGroup: Windows_NT
       archType: arm
+      platform: Windows_NT_arm
```
This is being fixed by https://github.com/dotnet/runtime/pull/642 also. 

Should we instead close that PR? We might want to run the tests @jashook mentioned there.

dagood replied:
About why it's here, I imagine we need the artifacts from the platform for this PR. (@jkoritzinsky can confirm.)

Are you referring to running the coreclr outerloop?

I think it would be easiest to drive the other PR to merge because it's more self-contained and probably easier to verify. Then the diff can be removed here.

jkoritzinsky replied:
I originally added this just so the job naming for the installer jobs was correct. I didn't realize that this was blocking the ARM testing at that time. We should be able to temporarily remove this and allow #642 to add it.

safern commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -0,0 +1,182 @@
+
+  - ${{ parameters.prepareSteps }}
+
+  - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
```
This could be moved to an `_IsOfficialBuild` common variable and we could add a variable template that is imported in runtime.yml that way we avoid duplication and these hard to read conditions. Something like: https://github.com/dotnet/runtime/blob/master/eng/pipelines/libraries/variables.yml#L9

dagood replied:
Interesting... I never got to grips with variable templates (or using variable values in template expressions), this is a good starting point. I'll give this a shot, thanks.

trylek replied:
Just please remember the variable templates can only be imported at the "top level" i.e. at the pipeline YAML level, not from the lower-level templates, that's the current limitation AFAIK. Like - it would be super-logical to import a variable template section into one of the CoreCLR template - it just doesn't work. I initially tried to mimic variables in templates by splitting platform-matrix in three different templates, then I rolled it back because it plain sucked. Santi currently worked it around by using explicit nested templates (the xplat-setup stuff). I believe this is just poor design on Azure Pipelines' part.

dagood replied:
Yeah... weird limits on templates. I don't see any fundamental reason you shouldn't be able to use them anywhere you want. (Whether it gets messy is up to the author!) Really confusing.

I guess this specific weirdness is because *pipeline* variables are "statically defined", but *job* variables aren't... but the only reference I find to statically defined variables is https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops so I can't tell if that's what's intended:

> In a parse time expression, you have access to `parameters` and statically-defined `variables`.

There's feedback filed on a few pages about this being confusing. 😛

safern replied:
Yeah. It would have to be imported in runtime.yml and then you can do `variables['IsOfficialBuild']`.... That's what we're doing for the libraries build now.

safern commented on eng/pipelines/installer/jobs/osx-build.yml:
```diff
@@ -1,38 +1,25 @@
-  - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
-    - task: NuGetAuthenticate@0
+    buildVariables:
+      CommonMSBuildArgs: >-
```
Just to learn, what does: `>-` do?

trylek replied:
I'm seeing this for the first time too...

dagood replied:
`>` replaces newlines with spaces (vs. `|` which keeps newlines), `-` strips the newline at the end. http://yaml-multiline.info/ is a nice resource for this, and it has links to the spec.

safern replied:
Thanks, pretty useful 😄

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -110,7 +110,19 @@ jobs:
               targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin/runtime
-          
+
+          - task: CopyFiles@2
```
I guess this will need to be rebased if: https://github.com/dotnet/runtime/pull/494 is merged first.

dagood replied:
Yeah, I think the MSBuild changes are probably going to be the bigger conflicts though. I'm not expecting to be able to merge them in particularly quick succession. 😄

safern commented on eng/pipelines/common/clone-checkout-bundle-step.yml:
```diff
@@ -0,0 +1,13 @@
+steps:
```
Should Libraries and CoreCLR jobs be updated to use this template? Should just be a couple of files: https://github.com/dotnet/runtime/search?q=Checkout_bundle&unscoped_q=Checkout_bundle

dagood replied:
Yeah, I initially waved off on it to avoid adding unnecessary changes while everyone's working on the yaml in various ways, but I'll give it a shot.

dagood committed:
Use clone-checkout-bundle-step.yml everywhere

dagood committed:
Try using isOfficialBuild var from template expr

dagood committed:
Switch from $[] to $() to fix clone template

$[] becomes empty string when a variable is undefined. $() sticks around literally if the variable doesn't exist, messing up the command. Unfortunately $[] doesn't seem to work in this context so we must use $() and define the variable as empty string.

dagood commented:
In the last CI run, I somehow got AzDO to give me a good runtime pipeline build of Installer using live bits. All the platforms ran but macOS, which was just stalling due to the ongoing AzDO issues.

@jkoritzinsky @ViktorHofer @safern @trylek PTAL

safern commented on eng/pipelines/common/variables.yml:
```diff
@@ -0,0 +1,3 @@
+variables:
```
I might be missing something, but where is this imported?

dagood replied:
Well, it's imported in Libraries' `eng\pipelines\libraries\variables.yml`. But yeah, not used by Installer yet (or anywhere else) because we don't have a def set up to do official builds yet. Shuffling this out of the CoreFX-specific vars just lets me switch `${{ and(ne(variables['System.TeamProject'], 'public'), notIn(variables['Build.Reason'], 'PullRequest')) }}` to `isOfficialBuild` in this PR with clear intent where it *will* come from once I work on an official build.

safern replied:
I see. I guess this file will be imported in runtime.yml for the official builds work under a global variable section just as `eng\pipelines\libraries\.azure-ci.yml` does for it's variables template.

We can leave that as-side for the official build effort.

dagood commented:
CI is all green except `runtime-coreclr (CoreCLR Pri0 Test Run R2R Windows_NT x86 checked)`, which timed out trying to send (get?) Helix results. I hit retry.

Would appreciate more reviews. 🙂

trylek reviewed (approved):
OMG, sorry for not signing off, your change looks awesome to me!

jkoritzinsky commented:
A lot of the Helix queues are very full (the Windows queue that PRs use currently has a queue depth of 2.3k work items), so we might want to wait for the queue to drain before re-running any more if it times out again.

jkoritzinsky commented on eng/pipelines/installer/installer-matrix.yml:
```diff
@@ -0,0 +1,100 @@
+    - Windows_NT_arm64
+    jobParameters:
+      ${{ insert }}: ${{ parameters }}
+      skipTests: true
```
It might be possible to make skipTests = crossBuild. Based on the current file, that seems correct.

jkoritzinsky commented on eng/pipelines/runtime.yml:
```diff
@@ -148,3 +148,16 @@ jobs:
+  parameters:
+    liveCoreClrBuildConfig: checked
+    liveLibrariesBuildConfig: Release
+    skipPlatforms:
```
I think we should invert this to instead specify the platforms we want to build for, not the platforms we want to skip. We should do this in a similar manner to the `platform-matrix.yml` file is done.

safern replied:
Anyway I think it makes sense to add a comment so that whenever we expand the full matrix for all partitions we don't forget adding any platform.

jkoritzinsky commented on src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets:
```diff
@@ -27,27 +30,45 @@
       <ReferenceCopyLocalPaths Remove="@(ReferenceCopyLocalPaths)" Condition="$([System.String]::Copy('%(FileName)').StartsWith('mscordaccore_'))" />
-      <ReferenceCopyLocalPaths Include="@(CoreCLRFiles)" />
+
+      <ReferenceCopyLocalPaths Include="@(CoreCLRFiles)" NuGetPackageId="$(MicrosoftNETCoreRuntimeCoreCLRPackage)" />
```
Why is the `NuGetPackageId` metadata required? Is that related to the support in Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk for crossgenning the framework?

dagood replied:
I believe it was a batch somewhere that assumed all items had `NuGetPackageId` defined. I wouldn't expect it to change any concrete results if the assumption were fixed instead, but this was easy and matches how we're adding this data to the CoreFX items.

jkoritzinsky commented on src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets:
```diff
@@ -124,11 +140,58 @@
+  </Target>
+
+  <!-- !!!BEGIN PATCHING Sharedfx SDK target -->
+  <Target Name="GetFilesFromPackageResolve">
```
What part of this target did you have to override/patch?

dagood replied:
See below:

```
      <!-- !!!BEGIN CHANGE Persist existing TargetPath. -->
      <RidSpecificFilesToPackage Condition="'%(RidSpecificFilesToPackage.TargetPath)' == ''">
      <!-- !!!END CHANGE -->
```

Although I guess I neglected mentioning the original: [`<RidSpecificFilesToPackage>`](https://github.com/dotnet/arcade/blob/d2d8d9f5c14e353acc7dfcf69c5fa5a8c368af40/src/Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk/targets/framework.dependency.targets#L20).

jkoritzinsky replied:
The paths without the change should work as-is. Were they not working correctly?

jkoritzinsky replied:
In fact, the paths should have been the same as what they're overridden to here.

dagood replied:
The cross-target paths get reset by that, and need to be preserved:

https://github.com/dotnet/runtime/blob/4e23cd3deabebf6c1f269379eeb01dbadc4be22e/src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets#L47-L49

I think they used to work by being added after `GetFilesFromPackageResolve` rather than before.

jkoritzinsky replied:
Good catch!

dagood replied:
Well, I did have the benefit of seeing the build fail before I did this. 😄 When the target paths are the same between cross and target archs, the zip gets duplicate entries which fails an unzip step later.

dagood commented on eng/pipelines/common/xplat-setup.yml:
```diff
@@ -39,7 +41,10 @@ jobs:
-      
+        # Set the bash script to display each command, and stop if any command exits nonzero.
+        - name: setScriptToEchoAndFailOnNonZero
+          value: 'set -xe'
```
We have them, it's just a no-op right now on Win a few lines up:

```
- name: setScriptToEchoAndFailOnNonZero
  value: ''
```

Don't know if it's possible to do this (or necessary) in batch.

dagood commented on eng/pipelines/installer/jobs/base-job.yml:
```diff
@@ -0,0 +1,173 @@
+    value: ''
+
+  - ${{ if ne(parameters.liveCoreClrBuildConfig, '') }}:
+    - name: liveCoreClrLegName
```
> I originally chickened out of Santi's suggestion to put this logic in xplat-setup (and you're not using it, after all)

I actually am using xplat-setup, since I use platform-matrix... or am I misunderstanding?

It makes sense to me to have it in xplat-setup, since the bulk of the transfer is setting up variables, and I think this is the only reasonable mechanism we have to share agent-local variable setup.

jkoritzinsky replied:
> > I originally chickened out of Santi's suggestion to put this logic in xplat-setup (and you're not using it, after all)
> 
> I actually am using xplat-setup, since I use platform-matrix... or am I misunderstanding?

We really need to update our docs on how these templates flow. I'll open a tracking issue for that.

safern replied:
The reasoning was to have all this `*ArtifactName` variables and `LegName` defined in `xplat-setup` so that all templates can benefit from it and just use them without having to actually define it in multiple places. Currently each partition defined this on it's base template.

However the change is not trivial as the right parameters need to be flown to the right templates in order to be able to access the values we need on `xplat-setup`. So this can be a follow up clean-up change. 

I still have to do clean-up on `libraries` template which I left from my previous PR to unblock work, so I can put it as part of my change.

trylek replied:
Please let's make sure we don't get deadlocked by these considerations for now - it's probably up to me to apologize for bringing them up in the first place. I'll be the happiest person in the world to clean this up but it's strictly ordered after standing up the official build and producing signing artifacts so please let's not block on that for now. I'll be happy to play with pipeline cleanups over / after the EOY holidays ;-).

dagood commented:
I don't think any of the convos are blockers--I'd like to get this merged to unblock and work on them incrementally. I've filed https://github.com/dotnet/runtime/issues/749 to track the issues. I'll file PRs for smaller ones really quick when I have gaps in my work on signing/publishing. Any objection (while that CoreCLR leg is finishing up)?
